src/hotspot/share/memory/metaspace/metachunk.hpp

```diff
@@ -100,9 +125,46 @@ class VirtualSpaceNode;
 //            |              |     "used"           |           |
 //            |              |                      |           |
 //            +--------------+ <- start   ----------+ ----------+
+//
+//
+// -- Relationships --
+//
+// Chunks are managed by a binary buddy style allocator
+//  (see https://en.wikipedia.org/wiki/Buddy_memory_allocation).
+// Chunks which are not a root chunk always have an adjoining buddy.
+//  The first chunk in a buddy pair is called the leader, the second
+//  one the follower.
+//
+// +----------+----------+
+// | leader   | follower |
+// +----------+----------+
+//
+//
+// -- Layout in address space --
+//
+// In order to implement buddy style allocation, we need an easy way to get
+//  from one chunk to the Metachunk representing the neighboring chunks
+//  (preceding resp. following it in memory).
+// But Metachunk headers and chunks are physically separated, and it is not
+//  possible to get the Metachunk* from the start of the chunk. Therefore
+//  Metachunk headers are part of a second linked list, describing the order
+//  in which their payload appears in memory:
+//
+// +---------+                       +---------+                       +---------+
+// |MetaChunk| <--next/prev_in_vs--> |MetaChunk| <--next/prev_in_vs--> |MetaChunk|
+// +---------+                       +---------+                       +---------+
+//     |                                 |                                  |
+//    base                              base                               base
+//     |                                 /                                  |
+//    /        --------------------------                                  /
+//   /        /          --------------------------------------------------
+//  |         |         /
+//  v         v         v
+// +---------+---------+-------------------+
+// |  chunk  |  chunk  |      chunk        |
+// +---------+---------+-------------------+
+//
```

Comment:
Do you have this same picture twice?  Other than that, thank you for the ascii art.  I think it'll help understand all of this code.

Reply:
Thought you like it :) The first picture describes the general division between headers and (possibly distant) chunks, the latter how the _prev_in_vs/_next_in_vs chains headers for adjacent chunks together. The latter is a result of some confusion Richard and I had during our last review round.